filecoin-proofs/src/api/post.rs

```diff
@@ -461,6 +644,199 @@ pub fn verify_winning_post<Tree: 'static + MerkleTreeTrait>(
     Ok(true)
 }
 
+// Partition a flat vector of vanilla proofs -- the post_config type
+// is required in order to determine the proper shape of the returned
+// partitioned proofs.
+pub fn partition_vanilla_proofs<Tree: MerkleTreeTrait>(
+    post_config: &PoStConfig,
+    pub_params: &fallback::PublicParams,
+    pub_inputs: &fallback::PublicInputs<<Tree::Hasher as Hasher>::Domain>,
+    partition_count: usize,
+    vanilla_proofs: &[FallbackPoStVanillaProof<Tree>],
+) -> Result<Vec<VanillaProof<Tree>>> {
+    info!("partition_vanilla_proofs:start");
+    ensure!(
+        post_config.typ == PoStType::Window || post_config.typ == PoStType::Winning,
+        "invalid post config type"
+    );
+
+    let num_sectors_per_chunk = pub_params.sector_count;
+    let num_sectors = pub_inputs.sectors.len();
+
+    ensure!(
+        num_sectors <= partition_count * num_sectors_per_chunk,
+        "cannot prove the provided number of sectors: {} > {} * {}",
+        num_sectors,
+        partition_count,
+        num_sectors_per_chunk,
+    );
+
+    let mut partition_proofs = Vec::new();
+
+    // Note that the partition proofs returned are shaped differently
+    // based on which type of PoSt is being considered.
+    if post_config.typ == PoStType::Window {
+        for (j, sectors_chunk) in pub_inputs.sectors.chunks(num_sectors_per_chunk).enumerate() {
+            trace!("processing partition {}", j);
+
+            let mut proofs = Vec::with_capacity(num_sectors_per_chunk);
+
+            for pub_sector in sectors_chunk.iter() {
+                let cur_proof = vanilla_proofs
+                    .iter()
+                    .find(|&proof| proof.sector_id == pub_sector.id)
+                    .expect("failed to locate sector proof");
+
+                // Note: Window post requires all inclusion proofs (based on the challenge
+                // count per sector) per sector proof.
+                proofs.extend(cur_proof.vanilla_proof.sectors.clone());
```

Comment:
This is confusing. It reads like `proofs` is a vector of sectors? Something is poorly named.